src/lib.rs

```diff
@@ -142,80 +135,112 @@ struct CoreSDK<WP> {
+                    }
+                    tokio::time::sleep(Duration::from_millis(100)).await;
+                }
+            };
+            let poll_result_future = $self.server_gateway.$gateway_fn($poll_arg);
+            tokio::select! {
+                _ = shutdownfut => {
+                    Err(CoreError::ShuttingDown)
+                }
+                r = poll_result_future => r.map_err(Into::into)
+            }
+        })
+    };
+}
+
 impl<WP> Core for CoreSDK<WP>
 where
     WP: ServerGatewayApis + Send + Sync + 'static,
 {
     #[instrument(skip(self), fields(pending_activation))]
-    fn poll_task(&self, task_queue: &str) -> Result<Task> {
-        // We must first check if there are pending workflow tasks for workflows that are currently
-        // replaying, and issue those tasks before bothering the server.
-        if let Some(pa) = self.pending_activations.pop() {
-            Span::current().record("pending_activation", &format!("{}", &pa).as_str());
-
-            let next_activation =
-                self.access_wf_machine(&pa.run_id, move |mgr| mgr.get_next_activation())?;
-            let task_token = pa.task_token.clone();
-            if next_activation.more_activations_needed {
-                self.pending_activations.push(pa);
-            }
-            return Ok(Task {
-                task_token,
-                variant: next_activation.activation.map(Into::into),
-            });
-        }
-
-        if self.shutdown_requested.load(Ordering::SeqCst) {
-            return Err(CoreError::ShuttingDown);
-        }
-
-        // This will block forever (unless interrupted by shutdown) in the event there is no work
-        // from the server
-        match self.poll_server(PollTaskRequest::Workflow(task_queue.to_owned())) {
-            Ok(PollTaskResponse::WorkflowTask(work)) => {
-                let task_token = work.task_token.clone();
-                debug!(
-                    task_token = %fmt_task_token(&task_token),
-                    "Received workflow task from server"
-                );
-
-                let (next_activation, run_id) = self.instantiate_or_update_workflow(work)?;
-
-                if next_activation.more_activations_needed {
-                    self.pending_activations.push(PendingActivation {
-                        run_id,
-                        task_token: task_token.clone(),
+    fn poll_task(&self, req: PollTaskReq) -> Result<Task> {
+        let task_queue = req.task_queue;
+        // The poll needs to be in a loop because we can't guarantee tail call optimization in Rust
+        // (simply) and we really, really need that for long-poll retries.
+        loop {
+            if req.workflows {
+                // We must first check if there are pending workflow tasks for workflows that are
+                // currently replaying, and issue those tasks before bothering the server.
+                if let Some(pa) = self.pending_activations.pop() {
```

Comment:
Can we pull checking for pending activations out of this loop?
I think we only need to loop for retrying the long poll, right?

Reply:
Not at all. Before this change it recursed to clear pending activations in the event of shutdown. The loop still serves that purpose